# 01. Import libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

# 02. Import data

In [2]:
# path folder
path = r'C:\\Users\alexp\Instacart Basket Analysis'
path

'C:\\\\Users\\alexp\\Instacart Basket Analysis'

In [3]:
# import df_merged_large.pkl
df_orders_products_large = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 03 If-Statements

## Q1 - Creating the "price_label" and "busiest_day" columns

In [4]:
# define a function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

## If the value within the “prices” column within the given row is less than or equal to 5, then return the string “Low-range product.”

## Or else, if the value within the “prices” column within the given row is greater than 5 and less than or equal to 15, then return the string “Mid-range product.”

## Or else, if the value of the “prices” column within the given row is greater than 15, then return the string “High-range product.”

## Or else, return the string “Not enough data.”

In [5]:
# apply the function using the loc() function and create the price_label column
df_orders_products_large.loc[df_orders_products_large['prices'] > 15, 'price_range_loc'] = 'High range product'

## If the values of the prices column of the df_orders_products_large dataframe are greater than 15, then return the string "High range product"

In [6]:
df_orders_products_large.loc[(df_orders_products_large['prices'] <= 15) & (df_orders_products_large['prices'] > 5), 'price_range_loc'] = 'Mid range product'

## If the values of the prices column of the df_orders_products_large_dataframe are less than or equal to 15 and greated than 5, then return the string "Mid range product"

In [7]:
df_orders_products_large.loc[df_orders_products_large['prices'] <= 5, 'price_range_loc'] = 'Low range product'

## If the values of the prices column of the df_orders_products_dataframe are less than 5, then return the string "Low range product"

In [8]:
df_orders_products_large['price_range_loc'].value_counts(dropna = False)

Mid range product     21860860
Low range product     10126321
High range product      417678
Name: price_range_loc, dtype: int64

In [9]:
df_orders_products_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  
0         Soda        77              7     9.0   both  Mid range product  
1         Soda        77              7     9.0   both  Mid range product  
2         Soda        77              7     9.0   both  Mid range product  
3         Soda        77              7     9.0   both  Mid range product  
4         Soda        77              7     9.0   both  Mid range product

In [10]:
# find the frequency of the orders_day_of_week
df_orders_products_large['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [11]:
# create the busiest day column using for-loops function
result = []

for value in df_orders_products_large["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [12]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [13]:
# create a new column within the orders_products_merged.pkl dataframe and set it equal to result
df_orders_products_large['busiest_day'] = result

In [14]:
# find the frequency of "busiest_day" column
df_orders_products_large['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

## Q2

In [15]:
# find the frequency of orders_day_of_week
df_orders_products_large['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

## 0 = Saturday (Busiest)
## 1 = Sunday (2nd busiest)
## 2 = Monday
## 3 = Tuesday (2nd slowest)
## 4 = Wednesday (Slowest)
## 5 = Thursday
## 6 = Friday

In [16]:
# create busiest days column using for-loops function
result = []

for value in df_orders_products_large["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest days")
  elif value == 1:
    result.append("Busiest days")
  elif value == 3:
    result.append("Slowest days")
  elif value == 4:
    result.append("Slowest days")
  else:
    result.append("Regular days")

In [17]:
result

['Regular days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Regular days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Slowest days

## Q3

In [18]:
# create a new column within the orders_products_merged.pkl dataframe and set it equal to result
df_orders_products_large['busiest_days'] = result

In [19]:
# find the frequency of the 'busiest days' column
df_orders_products_large['busiest_days'].value_counts(dropna = False)

Regular days    12916111
Busiest days    11864412
Slowest days     7624336
Name: busiest_days, dtype: int64

## Saturday (busiest) and Sunday (second busiest) have been combined and make a total of 11,864,412. In addition, Tuesday (second slowest) and Wednesday (slowest) are also combined and make a total of 7,624,336.

In [20]:
df_orders_products_large.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 17 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_week      int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
dtypes: category(1), float64(2), int64(10), object(4)
memory usage: 4.1+ GB


In [21]:
df_orders_products_large.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid range product   
1          Soda        77              7     9.0   both  Mid range product   
2          Soda        77              7     9.0   both  Mid range product   
3          Soda        77              7     9.0   both  Mid range product   
4          Soda        77              7     9.0   both  Mid range product   
5          Soda        77              7     9.0   both  Mid range product   
6          Soda        77              7     9.0   both  Mid range product   
7          Soda        77              7     9.0   both  Mid range product   
8          Soda        77              7     9.0   both  Mid range product   
9          Soda        77              7     9.0   both  Mid range produ

## Q4

In [22]:
# find the frequency of 'order_hour_of_day' column
df_orders_products_large['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

## Most orders: 9-17 is > 2000000
## Average orders: 8, 18-19 > 1000000
## Fewest orders: 0-7, 20-23 < 999999

In [23]:
# create busiest_period_of_day column using for-loops function
result = []

for value in df_orders_products_large["order_hour_of_day"]:
    if value in [10, 11, 14, 15, 13, 12, 16, 9, 17]:
        result.append("Most orders")
    elif value in [20, 7, 21, 22, 23, 6, 0, 1, 5, 2, 4, 3]:
        result.append("Fewest orders")
    else:
        result.append("Average orders")

In [24]:
result

['Average orders',
 'Fewest orders',
 'Most orders',
 'Fewest orders',
 'Most orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Most orders',
 '

In [25]:
df_orders_products_large['busiest_period_of_day'] = result

## Q5

In [26]:
# Print frequency of busiest_period_of_day column
df_orders_products_large['busiest_period_of_day'].value_counts(dropna = False)

Most orders       23205725
Average orders     4612925
Fewest orders      4586209
Name: busiest_period_of_day, dtype: int64

In [27]:
df_orders_products_large[['order_hour_of_day','busiest_period_of_day']].head(10)

order_hour_of_day busiest_period_of_day
0                  8        Average orders
1                  7         Fewest orders
2                 12           Most orders
3                  7         Fewest orders
4                 15           Most orders
5                  7         Fewest orders
6                  9           Most orders
7                 14           Most orders
8                 16           Most orders
9                  8        Average orders

In [28]:
df_orders_products_large.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid range product   
1         Soda        77              7     9.0   both  Mid range product   
2         Soda        77              7     9.0   both  Mid range product   
3         Soda        77              7     9.0   both  Mid range product   
4         Soda        77              7     9.0   both  Mid range product   
5         Soda        77              7     9.0   both  Mid range product   
6         Soda        77              7     9.0   both  Mid range product   
7         Soda        77              7     9.0   both  Mid range product   
8         Soda        77              7     9.0   both  Mid range product   
9         Soda        77              7     9.0   both  Mid range product   

      busiest_day  busiest_days busiest_period_of_day  
0  Regularly busy  Regular days        Average orders  
1  Regularly busy  Slowest days         Fewest orders  
2  Regularly busy  Slowest days           Most orders  
3      Least busy  Slowest days         Fewest orders  
4      Least busy  Slowest days           Most orders  
5  Regularly busy  Regular days         Fewest orders  
6  Regularly busy  Busiest days           Most orders  
7  Regularly busy  Busiest days           Most orders  
8  Regularly busy  Busiest days           Most orders  
9      Least busy  Slowest days        Average orders

## Q7

In [29]:
# Export updated df_orders_products_large to pkl
df_orders_products_large.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_columns.pkl'))